lib/orderbook/OrderBook.ts

```diff
@@ -1,130 +1,205 @@
+      const invoice: string = await this.getInvoice(order) as string;
+      const cast: OutgoingOrder = {
+        invoice,
+        id: order.id,
+        pairId: order.pairId,
+        price: order.price,
+        quantity: order.quantity,
+      };
+      outgoingOrders.push(cast);
+    }
+  }
+
+  private addOwnOrder = (order: orders.OwnOrder): matchingEngine.MatchingResult => {
     const matchingEngine = this.matchingEngines[order.pairId];
     if (!matchingEngine) {
       throw errors.INVALID_PAIR_ID(order.pairId);
     }
 
-    const stampedOrder: orders.StampedOwnOrder = { ...order, id: uuidv1(), createdAt: new Date() };
+    const stampedOrder: orders.StampedOwnOrder = { ...order, id: uuidv1(), createdAt: ms() };
     const matchingResult = matchingEngine.matchOrAddOwnOrder(stampedOrder);
     const { matches, remainingOrder } = matchingResult;
 
     if (matches.length > 0) {
-      const updatedOrders: Promise<void>[] = [];
-      for (const { maker, taker } of matches) {
+      matches.forEach(({ maker, taker }) => {
         this.handleMatch({ maker, taker });
-        updatedOrders.push(this.repository.updateOrderQuantity(maker.id, maker.quantity));
-      }
-      await Promise.all(updatedOrders);
+        this.updateOrderQuantity(this.ownOrders, maker, maker.quantity);
+      });
     }
     if (remainingOrder) {
       this.broadcastOrder(remainingOrder);
 
       if (this.config.internalmatching) {
-        const dbOrder = await this.repository.addOrder(<db.OrderFactory>remainingOrder);
-        this.logger.debug(`order added: ${JSON.stringify(dbOrder)}`);
+        this.addOrder(this.ownOrders, remainingOrder);
+        this.logger.debug(`order added: ${JSON.stringify(remainingOrder)}`);
       }
     }
 
     return matchingResult;
   }
 
-  private addPeerOrder = async (order: orders.PeerOrder): Promise<void> => {
+  private addPeerOrder = (order: orders.PeerOrder) => {
     const matchingEngine = this.matchingEngines[order.pairId];
     if (!matchingEngine) {
       this.logger.debug(`incoming peer order invalid pairId: ${order.pairId}`);
       return;
     }
 
-    const stampedOrder: orders.StampedPeerOrder = { ...order, createdAt: new Date() };
+    const stampedOrder: orders.StampedPeerOrder = { ...order, createdAt: ms() };
     matchingEngine.addPeerOrder(stampedOrder);
-    const dbOrder = await this.repository.addOrder(stampedOrder);
-    this.logger.debug(`order added: ${JSON.stringify(dbOrder)}`);
+    this.addOrder(this.peerOrders, stampedOrder);
+    this.logger.debug(`order added: ${JSON.stringify(stampedOrder)}`);
+  }
+
+  private updateOrderQuantity = (type: { [pairId: string]: Orders }, order: StampedOrder, decreasedQuantity: number) => {
+    const array = this.getOrderArray(type, order);
+    array.forEach((item, index) => {
+      if (item.id === order.id) {
+        item.quantity = item.quantity - decreasedQuantity;
+        if (item.quantity === 0) {
+          array.splice(index, 1);
+        }
```

Comment:
Although we might change the current design of keeping both `ownOrders` and `peerOrders` separated from `MatchingEngine`, I think we should already optimize it to avoid full-scan when querying on `orderId`. So instead using an array, just use a plain object, and use `orderId` for the keys.